dipy/align/mattes.pyx

```diff
@@ -0,0 +1,1390 @@
+                for i in range(nbins):
+                    mmarginal[j] += joint[i, j]
+
+
+cdef _joint_pdf_gradient_dense_2d(double[:] theta, Transform transform,
+                                  double[:,:] static, double[:,:] moving,
+                                  double[:,:] grid_to_space,
+                                  floating[:,:,:] mgradient, int[:,:] smask,
+                                  int[:,:] mmask, double smin, double sdelta,
+                                  double mmin, double mdelta, int nbins,
+                                  int padding, double[:,:,:] grad_pdf):
+    r''' Gradient of the joint PDF w.r.t. transform parameters theta
+
+    Computes the vector of partial derivatives of the joint histogram w.r.t.
+    each transformation parameter. The transformation itself is not necessary
+    to compute the gradient, but only its Jacobian.
+
+    Parameters
+    ----------
+    theta: array, shape (n,)
+        parameters of the transformation to compute the gradient from
+    transform: instance of Transform
+        the transformation with respect to whose parameters the gradient
+        must be computed
+    static: array, shape (R, C)
+        static image
+    moving: array, shape (R, C)
+        moving image
+    grid_to_space: array, shape (3, 3)
+        the grid-to-space transform associated to images static and moving
+        (we assume that both images have already been sampled at a common grid)
+    mgradient: array, shape (R, C, 2)
+        the gradient of the moving image
+    smask: array, shape (R, C)
+        mask of static object being registered (a binary array with 1's inside
+        the object of interest and 0's along the background)
+    mmask: array, shape (R, C)
+        mask of moving object being registered (a binary array with 1's inside
+        the object of interest and 0's along the background)
+    smin: float
+        the minimum observed intensity associated to the static image, which
+        was used to define the joint PDF
+    sdelta: float
+        bin size associated to the intensities of the static image
+    mmin: float
+        the minimum observed intensity associated to the moving image, which
+        was used to define the joint PDF
+    mdelta: float
+        bin size associated to the intensities of the moving image
+    nbins: int
+        number of histogram bins
+    padding: int
+        number of bins used as padding (the total bins used for padding at both
+        sides of the histogram is actually 2*padding)
+    grad_pdf: array, shape (nbins, nbins, len(theta))
+        the array to write the gradient to
+    '''
+    cdef:
+        int nrows = static.shape[0]
+        int ncols = static.shape[1]
+        int n = theta.shape[0]
+        int offset, valid_points, constant_jacobian=0
+        cnp.npy_intp k, i, j, r, c
+        double rn, cn
+        double val, spline_arg
+        double[:,:] J = np.ndarray(shape=(2, n), dtype=np.float64)
```

Comment:
Changed to np.empty() to be more explicit. I have found not-initializing the arrays useful to catch subtle bugs in my code (sometimes hidden by initializing to zero).